topostats/scars.py

```diff
@@ -0,0 +1,470 @@
+interpolated values filling the space of any detected scars."""
+import logging
+import numpy as np
+from topostats.logs.logs import LOGGER_NAME
+
+LOGGER = logging.getLogger(LOGGER_NAME)
+
+# pylint: disable=too-many-arguments
+# pylint: disable=too-many-instance-attributes
+
+
+class Scars:
+    """
+    A class for detecting and removing scars from an image. Scars are typically
+    long, thin horizontal streaks above/below the data, resulting from the
+    AFM imaging process.
+    """
+
+    def __init__(
+        self,
+        img: np.ndarray,
+        filename: str,
+        run: bool = True,
+        removal_iterations: int = 2,
+        threshold_low: float = 0.250,
+        threshold_high: float = 0.666,
+        max_scar_width: int = 4,
+        min_scar_length: int = 16,
+    ) -> None:
+        """
+        Initialise the class
+
+        Parameters
+        ----------
+        img: np.ndarray
+            A 2-D image to remove scars from.
+        filename: str
+            The filename (used for logging outputs only).
+        run: bool
+            A toggle to turn scar removal on, or off.
+        removal_iterations: int
+            The number of times the scar removal should run on the image.
+            Running just once sometimes isn't enough to remove some of the
+            more difficult to remove scars.
+        threshold_low: float
+            A value that when multiplied with the standard deviation, acts as a threshold to determine if an increase
+            or decrease in height might constitute the top or bottom of a scar.
+        threshold_high: float
+            A floating point value that is used similarly to threshold_low, however sharp inclines or descents
+            that result in values in the mask higher than this threshold are automatically considered scars.
+        max_scar_width: int
+            A value that dictates the maximum width that a scar can be. Note that this does not mean horizontal width,
+            rather vertical, this is because we conside scars to be laying flat, horizontally, so their width is
+            vertical and their length is horizontal.
+        min_scar_length: int
+            An integer that restricts the algorithm to only mark scars taht are as long or longer than this length.
+            This is important for ensuring that noise or legitimate but sharp datapoints do not get detected as scars.
+            Note that length here is horizontal, as scars are thin, horizontal features.
+
+        Returns
+        -------
+        None
+        """
+        self.img = img
+        self.filename = filename
+        self.run = run
+        self.removal_iterations = removal_iterations
+        self.threshold_low = threshold_low
+        self.threshold_high = threshold_high
+        self.max_scar_width = max_scar_width
+        self.min_scar_length = min_scar_length
+
+    def mark_if_positive_scar(
```

Comment:
A unit test would have an artificially created positive scar in isolation passed into this method and check to see if it is marked as a positive scar.

In addition a negative scar would also be passed in and the result tested to ensure its not marked as a positive scar.

It should be possible to write one test and paramterize the test.

Reply:
Unit tests added 👍